### ISSUE TITLE
Async version for remote_func.remote()

### ISSUE THREAD

**Issue by thedrow** [ray-project/ray#7229]
### Describe your feature request

Blocking while invoking remote functions is not acceptable if you have an event loop.
We should have an async implementation for invoking remote functions.

**Comment by simon-mo** [comment]
You can directly `await` on objectid or converting it to asyncio.Future now
https://ray.readthedocs.io/en/latest/async_api.html#objectids-as-asyncio-futures

**Comment by thedrow** [comment]
What if I don't want the result?
All I want is to invoke the remote function.

**Comment by simon-mo** [comment]
invoking remote function is non-blocking, `func.remote()` is fairly cheap and it immediately returns after invocation

**Comment by thedrow** [comment]
If it is non-blocking you may want to allow the user to context switch.
Otherwise if I wanted to rub `.remote()` in parallel to another operation I'd have to use a thread.

In addition, let's document that `.remote()` is non-blocking. Shell I open a separate issue on that?

**Comment by thedrow** [comment]
@simon-mo Ping?

**Comment by simon-mo** [comment]
Sorry about the delay. Can you suggest some use case where invoking remote task in parallel in a single process is required? 

Calling `func.remote` takes somewhere around 0.2-0.4 ms, this makes it possible to issue 200-500 remote calls per second from a single client per seconds. If you need to go beyond this, using several ray tasks to do this kind of submission isn’t a bad choice. 

Just calling `func.remote` doesn’t need to yield the event loop because we actually don’t do any network IO in this step. The “task submission” phase contains just serializing the arguments, building task specification, and handing it off to C++ threads.

**Comment by thedrow** [comment]
Now I understand.
So essentially the I/O will be done later asynchronously in a separate event loop with the GIL released.

I'd suggest we document this behavior. Shell I open another issue for it or do you have instructions for me now regarding a PR on the matter?

**Comment by richardliaw** [comment]
Can you provide a PR for improving the documentation (maybe in the docstring)? We'd really appreciate that!

**Comment by thedrow** [comment]
I'd specify this fact in the walkthrough as well.
It is important for people to know that they don't need to schedule `foo.remote()` in a thread.

**Comment by thedrow** [comment]
I do have a usecase where `foo.remote()` needs to block until the transmission is complete.
What if the cluster is unavailable? What if the cluster becomes unavailable during transmission?
How do I know that the transmission failed?

If a user wants to guarantee that the function will be invoked, the only way I can think of is to block until transmission is successful or an error occurs (in which case an exception will be raised).

**Comment by simon-mo** [comment]
We are working on revamping the signal API that will enables to you wait for some event user initiated events in the worker. 

Ideally:
```python
@ray.remote
def foo(signal):
    signal.send()

signal = RemoteSignal()
foo.remote(signal) # send the task to execute
ray.get(signal.wait())
```

You can use this implementation right now:
https://github.com/ray-project/ray/blob/f5b1062ed9834d7f485d0e47c397352b0e281066/python/ray/test_utils.py#L216-L237
